client/bloodhound/client.go

```diff
@@ -0,0 +1,399 @@
+package bloodhound
+
+import (
+	"bytes"
+	"compress/gzip"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"os"
+
+	"github.com/bloodhoundad/azurehound/v2/client/rest"
+	"github.com/bloodhoundad/azurehound/v2/constants"
+	"github.com/bloodhoundad/azurehound/v2/models"
+	"github.com/bloodhoundad/azurehound/v2/pipeline"
+	"github.com/go-logr/logr"
+)
+
+//go:generate go run go.uber.org/mock/mockgen -destination=./mocks/client.go -package=mocks . BloodHoundClient
+
+const (
+	BHEAuthSignature string = "bhesignature"
+)
+
+var ErrExceededRetryLimit = errors.New("exceeded max retry limit for ingest batch, proceeding with next batch...")
+
+// BloodHoundClient represents the methods for interacting with an instance of BloodHound
+type BloodHoundClient interface {
+	SendRequest(req *http.Request) (*http.Response, error)
+	CloseIdleConnections()
+	Ingest(ctx context.Context, in <-chan []interface{}) bool
+	GetAvailableJobs(ctx context.Context) ([]models.ClientJob, error)
+	Checkin(ctx context.Context) error
+	StartJob(ctx context.Context, jobId int) error
+	EndJob(ctx context.Context, status models.JobStatus, message string) error
+	UpdateClient(ctx context.Context) (*models.UpdateClientResponse, error)
+	EndOrphanedJob(ctx context.Context, updatedClient *models.UpdateClientResponse) error
+}
+
+// BHEClient implements the BloodHoundClient interface to communicate with a BloodHound Enterprise instance
+type BHEClient struct {
+	httpClient          *http.Client
+	bheUrl              url.URL
+	log                 logr.Logger
+	requestLimit        int
+	currentRequestCount int
+	maxRetries          int
+	retryDelay          int
+	proxy               string
+	token               string
+	tokenId             string
+}
+
+// NewBHEClient creates a new BloodHoundClient using the values from the application's config
+func NewBHEClient(bheUrl url.URL, tokenId, token, proxy string, maxReqPerConn, maxRetries int, logger logr.Logger) (BloodHoundClient, error) {
+	client, err := rest.NewHTTPClient(proxy)
+	if err != nil {
+		return nil, err
+	}
+
+	client.Transport = signingTransport{
+		base:      client.Transport,
+		tokenId:   tokenId,
+		token:     token,
+		signature: BHEAuthSignature,
+	}
+
+	return &BHEClient{
+		httpClient:          client,
+		bheUrl:              bheUrl,
+		requestLimit:        maxReqPerConn,
+		currentRequestCount: 0,
+		maxRetries:          maxRetries,
+		proxy:               proxy,
+		retryDelay:          5,
+		log:                 logger,
+	}, nil
+}
+
+// SendRequest sends a given request to the BHE instance. In the event of an error, 3 retries will be attempted
+func (s *BHEClient) SendRequest(req *http.Request) (*http.Response, error) {
+	var (
+		res *http.Response
+	)
+
+	// copy the bytes in case we need to retry the request
+	if body, err := rest.CopyBody(req); err != nil {
+		return nil, err
+	} else {
+		for currentAttept := 0; currentAttept <= s.maxRetries; currentAttept++ {
```

Comment:
I renamed retry to currentAttempt as it made more sense
Since this logic is guaranteed to run at least one time before any retry can occur, it made more sense to me to rename this